content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md

```diff
@@ -0,0 +1,864 @@
+
+In this section, I will demonstrate the manual installation of crun with the WasmEdge runtime library, and adjust containerd config to use crun as the runtime, enabling Wasm support on the Kubernetes node.
+
+Create a single-node Kubernetes cluster using kind.
+
+```bash
+kind create cluster --name wasm-demo
+```
+
+Each Kubernetes node created by kind is a Docker container, you can enter that node using the `docker exec` command.
+
+```bash
+docker exec -it wasm-demo-control-plane bash
+```
+
+{{< note >}}
+After entering the node, follow the instructions in the section: [Low-level Container Runtime](#low-level-container-runtime) to install crun binary on the node.
+{{< /note >}}
+
+Modify the containerd configuration file `/etc/containerd/config.toml`, add the following content at the end:
+- Configure `crun` as the runtime handler for containerd. The format is `[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.${HANDLER_NAME}]`.
+- `pod_annotations` allows passing the annotation `module.wasm.image/variant` to crun, which is set in Pod metadata to identify the Wasm workload. 
+
+```bash
+cat >> /etc/containerd/config.toml << EOF
+[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.crun]
+    runtime_type = "io.containerd.runc.v2"
+    pod_annotations = ["module.wasm.image/variant"]
+[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.crun.options]
+    BinaryName = "crun"
+EOF
+```
+
+Restart containerd.
+
+```bash
+systemctl restart containerd
+```
+
+Add label `runtime=crun` on the node.
+
+```bash
+kubectl label nodes wasm-demo-control-plane runtime=crun
+```
+
+Create a RuntimeClass resource named `crun` to use the pre-configured `crun` handler in containerd, the `scheduling.nodeSelector` property sends pod to nodes with the `runtime=crun` label.
+
+```yaml
+apiVersion: node.k8s.io/v1
+kind: RuntimeClass
+metadata:
+  name: crun
+scheduling:
+  nodeSelector:
+    runtime: crun
+handler: crun
+```
+
+Set `.spec.runtimeClassName` for the pod to target the pod at the `crun` RuntimeClass. This will ensure the pod gets assigned to a node and runtime specified in the `crun` RuntimeClass. Additionally, set the annotation `module.wasm.image/variant: compat` to inform crun that this is a Wasm workload.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: wasm-demo-app
+  annotations:
+    module.wasm.image/variant: compat
+spec:
+  runtimeClassName: crun
+  containers:
+  - name: wasm-demo-app
+    image: docker.io/cr7258/wasm-demo-app:v1
+```
+
+You can use `port-forward` to forward traffics to your local machine for access.
+
+```bash
+kubectl port-forward pod/wasm-demo-app 8080:8080
+```
+
+Open a new terminal, send a request to the service.
+
+```bash
+curl http://localhost:8080
+```
+
+The output is:
+
+```bash
+Hello, World!
+```
+
+Once the testing is complete, you can destroy the cluster using this command.
+
+```bash
+kind delete cluster --name wasm-demo
+```
+##### KWasm Operator
```

Comment:
This could be:
```markdown
##### Set up the KWasm Operator
```
or
```suggestion
##### Set up your cluster for Wasm using KWasm, then run the app inside a pod
```